usdShade: Adding NormalMapTexture validator

beersandrew opened the pull request:
### Description of Change(s)
Added an implementation of NormalMapTexture validation rules 

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3?pane=issue&itemId=65425006

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: initial implementation

beersandrew committed:
feat: finish normalMapTexture implementation and tests

- added new validatorTokens
- added implementation to match previous complianceChecker implementation
- added tests to test each failure condition

beersandrew committed:
fix: adjust include path

beersandrew committed:
fix: update validatorToken names

beersandrew commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -590,6 +589,244 @@ _SubsetsMaterialBindFamily(const UsdPrim& usdPrim)
+    }
+
+    const UsdShadeShader shader(usdPrim);
+    if (!shader) {
```
@tallytalwar I see this check here: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/complianceChecker.py#L468 however I'm not sure what USD structure would trigger this failure, after it's already passed the `IsA` check.

tallytalwar replied:
Yeah I don't think so also.

beersandrew commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -590,6 +589,244 @@ _SubsetsMaterialBindFamily(const UsdPrim& usdPrim)
+    {
+        errors.emplace_back(
+            UsdShadeValidationErrorNameTokens->nonCompliantScale,
+            UsdValidationErrorType::Warn,
```
@tallytalwar in the compliance checker, this is a warning, however the bias check below is an error. Is that correct? I wanted to change the bias error below to be a warning, let me know what you think

tallytalwar replied:
Referring the original commit for bias and scale adjustments to compliance checker: https://github.com/PixarAnimationStudios/OpenUSD/commit/b718721744f72c310f9b08aeac101c9fecd9d89a

When complying with UsdPreviewTexture and UsdUVTexture (that is when nonCompliantScaleValues is false) we need to make sure to report error for bias values. Refer these here: https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usdImaging/plugin/usdShaders/shaders/shaderDefs.usda#L100-L112

Regarding non-conforming scale values being reported as warnings: There is a reasoning here: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/complianceChecker.py#L550-L552 (I will request to keep all these code comments as-is in the cpp so as to save the reasoning). I believe Blender is the application mentioned here (cc. @dgovil is that still the case?)

beersandrew replied:
made sure to include all comments from the complianceChecker code (I did miss one). Also rebased this branch

beersandrew commented on pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp:
```diff
@@ -557,6 +559,213 @@ TestUsdShadeEncapsulationRulesValidator()
+}
+
+void
+TestUsdShadeNormalMapTextureValidator()
```
@tallytalwar There are quite a few failure conditions in this test and validator, let me know if you think it would be better to break this up in to multiple validators or not

beersandrew commented on pxr/usd/usdShade/testenv/testUsdShadeValidators/normalMap.jpg:
@tallytalwar We had the issue in the past with the encapsulation validator where me adding external files did not work for you, just calling this out here as this may happen again as I don't know what the difference was here in the past.

jesschimein commented:
Filed as internal issue #USD-10296

jesschimein commented:
/AzurePipelines run

beersandrew committed:
fix: fixing various rebasing issues

beersandrew committed:
fix: addressing comments

- remove unnecessary check
- add comment from complianceChecker.py that was missed

jesschimein commented:
/AzurePipelines run

beersandrew commented:
Closing this PR, this work is now done on https://github.com/PixarAnimationStudios/OpenUSD/pull/3443
